functions/grapher/[slug].ts

```diff
@@ -41,7 +44,49 @@ export const onRequestGet: PagesFunction = async (context) => {
             { ...env, url },
             request.headers.get("if-none-match")
         )
-        .catch((e) => error(500, e))
+        .catch(async (e) => {
+            // Here we do a unified after the fact handling of 404s to check
+            // if we have a redirect in the _grapherRedirects.json file.
+            // This is done as a catch handler that checks for 404 pages
+            // so that the common, happy path does not have to fetch the redirects file.
+            if (e instanceof StatusError && e.status === 404) {
+                console.log("Handling 404 for ", url.pathname)
+                const fullslug = url.pathname.split("/").pop() as string
+
+                const allExtensions = Object.values(extensions)
+                    .map((ext) => ext.replace(".", "\\.")) // for the regex make sure we match only a single dot, not any character
+                    .join("|")
+                const regexForKnownExtensions = new RegExp(
+                    `^(?<slug>.*?)(?<extension>${allExtensions})?$`
+                )
+
+                const matchResult = fullslug.match(regexForKnownExtensions)
+                const slug = matchResult?.groups?.slug ?? fullslug
+                const extension = matchResult?.groups?.extension ?? ""
```

Comment:
So, what this code doesn't currently seem to handle is non-lowercase slugs.
Our rule is that a slug needs to be all-lowercase, and if it isn't we want to redirect it.

In the old code, I tried to be clever about it so we never need to run through two redirects, avoiding the scenario `REDIR-SOURCE -> redir-source -> redir-target`. I don't think we need to take care of that again, which means we would just do:

```ts
if (slug.toLowerCase() !== slug)
  return redir(302, slug.toLowerCase())
```